io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java

```diff
@@ -406,12 +448,27 @@ private void proceedOrClose() {
 
 		private void createNextIterator() {
 			URI nextUri;
+			URIBuilder builder;
 			try {
 				nextUri = calculateNextUri();
+				builder = new URIBuilder(nextUri);
 			} catch (URISyntaxException e) {
 				throw new IllegalStateException(e.getMessage(), e);
 			}
 
+			// Download all the files into a folder for Bucket service
+			if (tmpdir != null) {
+				try {
+					URL url = nextUri.toURL();
+					String fileName = tmpdir + url.getPath() + UUID.randomUUID();
+					// TODO: may be download files in a thread
+					FileUtils.copyURLToFile(url, new File(fileName));
```

Comment:
To avoid downloading the same data twice I think it would make sense to overwrite `nextUri` with the URI to the downloaded file.